fabric-fluid-v1/src/main/java/net/fabricmc/fabric/api/fluid/v1/world/SidedFluidContainer.java

```diff
@@ -0,0 +1,66 @@
+package net.fabricmc.fabric.api.fluid.v1.world;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.block.FluidDrainable;
+import net.minecraft.block.FluidFillable;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.BlockView;
+import net.minecraft.world.World;
+import net.minecraft.world.WorldAccess;
+
+import net.fabricmc.fabric.api.fluid.v1.Action;
+import net.fabricmc.fabric.api.fluid.v1.container.FluidContainer;
+import net.fabricmc.fabric.api.fluid.v1.container.volume.FluidVolume;
+import net.fabricmc.fabric.api.fluid.v1.container.volume.SimpleFluidVolume;
+import net.fabricmc.fabric.api.fluid.v1.math.Drops;
+
+/**
+ * A container with different fluid containers per face, can be implemented on entities, and blocks (does not check block entities!).
+ */
+public interface SidedFluidContainer extends FluidFillable, FluidDrainable {
+	@Override
+	default Fluid tryDrainFluid(WorldAccess world, BlockPos pos, BlockState state) {
+		if (world instanceof World) {
+			FluidVolume volume = this.getContainer((World) world, pos, null).drain(Drops.getBucket(), Action.SIMULATE);
+
+			if (volume.getFluid() != Fluids.EMPTY && volume.getAmount() == Drops.getBucket()) {
+				return this.getContainer((World) world, pos, null).drain(Drops.getBucket(), Action.PERFORM).getFluid();
+			}
+		}
+
+		return Fluids.EMPTY;
+	}
+
+	/**
+	 * Get the container for the given face.
+	 *
+	 * @param world the world
+	 * @param pos the position
+	 * @param face the face being accessed, may be null
+	 * @return the container
+	 */
+	FluidContainer getContainer(World world, BlockPos pos, Direction face);
+
+	@Override
+	default boolean canFillWithFluid(BlockView world, BlockPos pos, BlockState state, Fluid fluid) {
+		if (world instanceof World) {
+			return this.getContainer((World) world, pos, null).consume(new SimpleFluidVolume(fluid, Drops.getBucket()), Action.SIMULATE).getAmount() == Drops.getBucket();
+		}
+
+		return false;
+	}
+
+	@Override
+	default boolean tryFillWithFluid(WorldAccess world, BlockPos pos, BlockState state, FluidState fluidState) {
+		if (this.canFillWithFluid(world, pos, state, fluidState.getFluid())) {
+			this.getContainer((World) world, pos, null).consume(new SimpleFluidVolume(fluidState.getFluid(), Drops.getBucket()), Action.PERFORM);
```

Comment:
might just accept the `World` parameter so you don't have to cast and instanceof

Reply:
this is a mojang method